node_modules/oae-principals/lib/api.user.js

```diff
@@ -614,6 +615,128 @@ var deleteUser = module.exports.deleteUser = function(ctx, userId, callback) {
 };
 
 /**
+ * 
+ * Delete or restore users within a tenancy
+ * 
+ * @function deleteOrRestoreUsersByTenancy
+ * @param  {String}     tenantAlias      The tenant alias we want to delete or restore users from
+ * @param  {Boolean}    disableUsers     Sets the behaviour to delete if true and restore if false
+ * @param  {Function}   callback         Standard callback function
+ * @param  {Object}     callback.err     An error that occured, if any
+ * @param  {Object[]}   callback.users   An array of objects representing affected users
+ */
+var deleteOrRestoreUsersByTenancy = module.exports.deleteOrRestoreUsersByTenancy = function (tenantAlias, disableUsers, callback) {
+    _getAllUsersForTenant(tenantAlias, function(err, users) {
+        if (err) {
+            callback(err);
+        }
+
+        if (disableUsers) {
+            _deletePrincipals(users, function(err, users) {
+                if (err) {
+                    callback(err);
+                }
+                callback(null, users);
+            });
+        } else {
+            _restorePrincipals(users, function (err, users) {
+                 if (err) {
+                    callback(err);
+                }
+                callback(null, users);
+            });
+        }
+    });
+};
+
+/**
+ * Page through all the users in the system, and filter them by tenant
+ *
+ * @param  {String}     tenantAlias             The tenant alias to filter by
+ * @param  {Function}   callback                Invoked when users have been collected
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {Object[]}   callback.userHashes     An array of user hashes
+ * @api private
+ */
+var _getAllUsersForTenant = function (tenantAlias, callback) {
```

Comment:
We desperately need a better way of collecting all users for a tenant, this is too slow.